### PR TITLE
Static code hooks at compile time

### PR DESCRIPTION
**Pull request by kripken**
This lets us do stuff like this:
```
-  $PIPEFS__postset: '__ATINIT__.push(function() { PIPEFS.root = FS.mount(PIPEFS, {}, null); });',
+  $PIPEFS__postset: function() {
+   addStaticAtInit('PIPEFS.root = FS.mount(PIPEFS, {}, null);');
+  },
```
The code then appears in the right place for at-initialization time, in both minimal and normal runtimes. And then we can remove the `postamble_minimal.js` hack with hardcoded initializers.

There are also hooks for at-main and at-exit. For the minimal runtime, calling those does nothing currently, as the minimal runtime code doesn't have the placeholder to receive those. If we decide we want to support those hooks there, it's easy to add later.

This PR uses those hooks in the relevant library code. Later work can implement LLVM static constructors using this mechanism too, which would allow us to simplify things even more, but I didn't want to do too much in one PR.

To enable this PR, I also made it possible for a library `__postset` to be a function that is called at compile time.

### CODE REVIEW

**Comment by kripken** [comment]
ping @juj

File: src/utility.js [hunk 1]
```diff
@@ -179,41 +179,6 @@ function splitter(array, filter) {
...
-
-var _PROF_ORIGIN = Date.now();
-var _PROF_TIME = _PROF_ORIGIN;
-function PROF(pass) {
-  if (!pass) {
-    dprint("Profiling: " + ((Date.now() - _PROF_TIME)/1000) + ' seconds, total: ' + ((Date.now() - _PROF_ORIGIN)/1000));
-  }
-  PROF_TIME = Date.now();
-}
-
```
Comment (juj) [inline src/utility.js]: This was removing just old dead code?

File: src/library_fs.js [hunk 2]
```diff
@@ -21,14 +21,15 @@ mergeInto(LibraryManager.library, {
...
-                '__ATMAIN__.push(function() { FS.ignorePermissions = false });' +
-                '__ATEXIT__.push(function() { FS.quit() });' +
-#endif
-                // Get module methods from settings
-                '{{{ EXPORTED_RUNTIME_METHODS.filter(function(func) { return func.substr(0, 3) === 'FS_' }).map(function(func){return 'Module["' + func + '"] = FS.' + func.substr(3) + ";"}).reduce(function(str, func){return str + func;}, '') }}}',
+  $FS__postset: function() {
+    // TODO: do we need noFSInit?
+    addStaticAtInit('if (!Module["noFSInit"] && !FS.init.initialized) FS.init();');
+    addStaticAtMain('FS.ignorePermissions = false;');
+    addStaticAtExit('FS.quit();');
```
Comment (juj) [inline src/library_fs.js]: I wonder if saying `addStaticAtInit()` is particularly illustrative, note how there's `FS.staticInit();` also below, with two different meanings of static(?). Perhaps `addRuntimeAtInit()` or just `addAtInit()`? Even though the registration is evaluated statically/at compile time, there is nothing much static about these - the functions are still run at runtime, and not at compile time.

(`addStaticAtInit()` is also ok if no better name)

**Comment by juj** [comment]
lgtm!

File: src/utility.js (hunk 1 above)
Comment (kripken) [reply src/utility.js]: Yeah, this old profiler code has not been useful since the asm.js backend, I believe.

File: src/library_fs.js (hunk 2 above)
Comment (kripken) [reply src/library_fs.js]: Hmm, yeah, maybe "static" means sometihng different in different cases. I intended static to mean we decide about it at compile time. But yeah, we run it at runtime. So maybe it's not the best name. Changing to "addAtInit", thanks.

**Pull request merged**